### ISSUE TITLE
Accessing other rows than the current row

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#200]
_Original author: sanderse...@gmail.com (November 13, 2010 22:18:28)_

I've searched through the issues and all the documentation I could find and come up blank, sorry if this has been answered elsewhere.

Is there any way of accessing rows other than the current row? Something like rows[rowIndex-1].cells[&quot;Column1&quot;].value would be priceless. Unfortunately it doesn't work.

I'm looking for ways of computing moving averages etc and I need this feature. Thanks, and keep up the good work, this software looks great!

_Original issue: http://code.google.com/p/google-refine/issues/detail?id=200_

**Comment by wetneb** [comment]
For use cases like this, where it is not clear how to achieve the desired result via the records mode, I would be in favour of letting users run code (Python, R…) on the entire table, modifying it in arbitrary ways. It would avoid the need to export the project, run it through a script or some other tool, and import it back again.

We would need to expose the entire table in a suitable object in the guest language: pandas dataframe in Python, table in R…

**Comment by thadguidry** [comment]
@wetneb or Apache Drill ;-)

**Comment by wetneb** [comment]
Drill is a query engine so it's less clear to me what its integration as an operation would look like - would you use it to replace the table by the query results? Or would you try to add support for `INSERT` statements which modify the table? It'd be good to have concrete workflow examples to make sure we are on the same page.

(But I agree their integration of structured data in a columnar environment is clearly related)

**Comment by thadguidry** [comment]
Supporting Selections (similar to how we allow users to select subdata various ways with GREL functions now, via parseJson() and parseHtml(), etc.)

I initially thought of "wouldn't it be cool if" somehow exposing a new `select()` and letting Apache Drill handling the heavylifting?  It would be useful to know how others view this issue however like @ettorerizza

`select()` could be exposed as a Data Model transform function via an inverse `insert()`.
And it doesn't have to be just at a GREL level...Our Expression editor UI could even be enhanced for support of this in a better fashion. (given time/money)

**Comment by antoine2711** [comment]
> Is there any way of accessing rows other than the current row? Something like rows[rowIndex-1].cells["Column1"].value would be priceless. Unfortunately it doesn't work.

Hi @tfmorris, while reading this, it came to me that now that the cross() function has been enhanced, I think you could get `rows[rowIndex-1].cells["Column1"].value` this way:

1. Have a numerical key column where `value = row.index + 1`;
2. where you want the value from another row, just put: `cross(_iID_, "PRJ-NAME", "COL-NAME").[0].cells.value`. Note that you can even get multiple values from this row in the same call: `with(cross("iID", "PRJ-NAME", "COL-NAME").[0].cells), vR, vR["ID"] + vR["Name"] + vR["URL"])` where _iID_ is of course the numerical value, and _PRJ-NAME_ and _COL-NAME_ the name of the joined table and it's key column.

Is there an angle I'm not seeing?

Regards, Antoine

**Comment by tfmorris** [comment]
@antoine2711 Anything that you see which was "authored" by me on October 14, 2012 is usually just an artifact of the import that I did of all old issues from Google Code on that date.

You can see the original author (with obfuscated email) and date here:

> Original author: sanderse...@gmail.com (November 13, 2010 22:18:28)

Your workaround sounds plausible on the face of it, although I haven't tested it. Syntactically it's quite a bit more unwieldy than something like `rows[rowIndex-1]`.

**Comment by antoine2711** [comment]
> @antoine2711 Anything that you see which was "authored" by me on October 14, 2012 is usually just an artifact of the import that I did of all old issues from Google Code on that date.

Shame, @tfmorris, it would have been a very productive day… 3:-)

Yes, `rows[rowIndex-1]` is not as elegant as `cross(_iID_, "PRJ-NAME", "COL-NAME")[0]`, still, it much more flexible as it has not project boundary, and the point I was making is that it now exists in v3.4 and it would answer the need.

**Comment by antoine2711** [comment]
> Syntactically it's quite a bit more unwieldy than something like `rows[rowIndex-1]`.

Yes, but, this made me think, and I created this issue for improvement: Enhance `cross()` with empty "project name" calling parameter to represent the current project and empty column name to represent the index #2504. 

Now, if that gets done, the comparaison would be:

`rows[rowIndex-1]` vs `cross(rowIndex-1, "", "")[0]` or even this `cross(rowIndex-1)[0]`

The difference would be cosmetic.

Regards, Antoine

**Comment by antoine2711** [comment]
@tfmorris: With PR #2518, what is asked here will be doable like this:

`cross(rowIndex-1)[0]`

In essence, it would do EXACTLY this: `rows[rowIndex-1]`, if it existed.

Regards, Antoine
